var/spack/repos/builtin/packages/fmt/package.py

```diff
@@ -35,7 +35,7 @@ class Fmt(CMakePackage):
             values=('98', '11', '14', '17'),
             multi=False,
             description='Use the specified C++ standard when building')
-    variant('pic', default=True, description='Enable generation of position-independent code')
+    variant('shared', default=False, description='Build shared library')
```

Comment:
Most packages seem to default to `True`, which makes sense in my opinion. Any reason we want to disable shared libraries by default?

Reply:
For Debian, the Debian maintainer explained it this way (in 2017):

> https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=878070#10
> fmtlib changed its major version 4 times in the last 2½ years, so considering its small size and relative unstability (so far) the package doesn't provide a shared library right now. In version 4 there are less breaking changes than before, so I'll re-evaluate whether to add a shared library later in the release cycle.

The issue which this commit introduces is, that it removes the pic variant which added `self.compiler.cxx_pic_flag`:

> Reference: https://fmt.dev/latest/usage.html#building-the-library
> To build a static library with position independent code (required if the main consumer of the fmt library is a shared library i.e. a Python extension) set the CMAKE_POSITION_INDEPENDENT_CODE CMake variable to TRUE:
> `cmake -DCMAKE_POSITION_INDEPENDENT_CODE=TRUE ...`

Currently, by default, `fmt` is built as `fmt+pic`, meaning it is compiled as position-independent code similar to the above, but we let our `self.compiler` object specify which PIC compiler flag to use. This can be beneficial for architectures where there are small and large position offsets and where the the small-offset-position-independent code is much faster than the large-offset-position-independent code. I know that S/390 is such an architecture (which is not of any relevance nowadays), but there may be others now or later.

For example, with `gcc`, this adds `-fPIC` to the `CFLAGS` and `CXXFLAGS` by default, but some architectures, code with '-fpic' can be faster. For x86_64 this makes no difference, I assume only older 32-bit architectures might actually differentiate between -fpic and -fPIC, but other compilers or architectures might have other considerations.

Your current branch, by removing the `self.compiler.cxx_pic_flag` from `CMAKE_CXX_FLAGS` changes this compile the static library as **position-_dep_endent-code** by default!!

This is the code we need to have, which this PR currently removes
```py
        if '+pic' in spec:
            args.extend([
                '-DCMAKE_C_FLAGS={0}'.format(self.compiler.cc_pic_flag),
                '-DCMAKE_CXX_FLAGS={0}'.format(self.compiler.cxx_pic_flag)
            ])
```

A python extension using `libfmt.a` would be just one example possibly needing this. `cracklib` is another one, which is a shared library packaged by spack, and spack users might have their own shared libraries using fmt.

The consequence of compiling libfmt as position-dependent-code is that the assembler has to set the DT_TEXTREL flag in the dynamic section of the assembled object file and it has to add relocation information for those code and read-only data locations which are compiled to need absolute references. At runtime, the dynamic linker then has to remap these read-only code and data sections read-write and patch the contained absolute references to the real absolute references before execution of the program.

This then defeats the code-sharing property of shared libraries in memory: By patching these read-only pages, the kernel has to create read-write copies of these pages in memory. If large libraries use libfmt, the whole library is affected by this and the memory consumption many parallel users of the same library multiplies for each use of the library in a program.

In addition, writing to read-only code and data sections of a library is usually not allowed by SELinux and would have to be explicitly allowed.

Therefore, the current default of compiling also the static libfmt.a as PIC makes sense and I would not remove it.
Also, I would not remove the default static library libfmt,a as that could break builds of spack customers which explicitly request the static build of libfmt when building their projects. Because of this impact, I'd vote to keep the static library by default.

Reply:
Looking at Homebrew's formula, one could imagine it equivalent to:
https://github.com/Homebrew/homebrew-core/blob/master/Formula/fmt.rb#L19
```py
    system "cmake", ".", "-DBUILD_SHARED_LIBS=FALSE", *std_cmake_args
    system "make"
    system "make", "install"
  if  build-shared
    system "cmake", ".", "-DBUILD_SHARED_LIBS=TRUE", *std_cmake_args
    system "make", "install"
    system "make", "clean"
```
But still to not break assumptions that our libfmt is by default only a static lib, I'd not do the shard lib build by default.

And, I have strong doubts what the benefits of a shared libfmt actually are: An optimized libfmt is only 229k in size, so there is not much to gain from making such a small library shared by default.

Reply:
Thanks for the detailed explanation! So I guess what we should do is:
1. Make the code position-independent by default (using `-DCMAKE_POSITION_INDEPENDENT_CODE=TRUE`)
2. Install the shared library *in addition* to the static library

Reply:
Yeah, but in all honesty, I would not change a single line of code implementing the `pic` variant of `fmt`:

Unless all versions of `cmake` handle `-DCMAKE_POSITION_INDEPENDENT_CODE=TRUE` the same as we do currently by adding `-DCMAKE_CXX_FLAGS={0}'.format(self.compiler.cxx_pic_flag)` meaning all cmake versions know this about all compilers spack supports, we could only add regressions for users of these other compilers.

For gcc, clang or so, of course cmake knows, to add `-fPIC` but for other compilers, I am not sure. Therefore I would not risk regressing a state which is not broken and works now:

`self.compiler.cxx_pic_flag` works with any version of `cmake` and even for a custom compiler which cmake can't know about.